test/agents/test_agent.py

```diff
@@ -0,0 +1,249 @@
+        )
+    )
+    assert len(agent.tools) == 1
+    assert "Retriever" in agent.tools
+    assert agent.has_tool(tool_name="Retriever")
+    assert isinstance(agent.tools["Retriever"].pipeline_or_node, BaseComponent)
+
+    agent.add_tool(
+        Tool(
+            name="Retriever",
+            pipeline_or_node=retriever,
+            description="useful for when you need to retrieve documents from your index",
+        )
+    )
+
+    # Add a Pipeline as a Tool to an Agent and overwrite the previously added Tool
+    retriever_pipeline = DocumentSearchPipeline(MockRetriever())
+    agent.add_tool(
+        Tool(
+            name="Retriever",
+            pipeline_or_node=retriever_pipeline,
+            description="useful for when you need to retrieve documents from your index",
+        )
+    )
+    assert len(agent.tools) == 1
+    assert "Retriever" in agent.tools
+    assert agent.has_tool(tool_name="Retriever")
+    assert isinstance(agent.tools["Retriever"].pipeline_or_node, BaseStandardPipeline)
+
+
+def test_agent_chooses_no_action():
+    agent = Agent(prompt_node=MockPromptNode())
+    retriever = MockRetriever()
+    agent.add_tool(
+        Tool(
+            name="Retriever",
+            pipeline_or_node=retriever,
+            description="useful for when you need to retrieve documents from your index",
+        )
+    )
+    with pytest.raises(AgentError, match=r"Wrong output format.*"):
+        agent.run("How many letters does the name of the town where Christelle lives have?")
+
+
+def test_max_iterations(caplog, monkeypatch):
+    # Run an Agent and stop because max_iterations is reached
+    agent = Agent(prompt_node=MockPromptNode(), max_iterations=1)
+    retriever = MockRetriever()
+    agent.add_tool(
+        Tool(
+            name="Retriever",
+            pipeline_or_node=retriever,
+            description="useful for when you need to retrieve documents from your index",
+        )
+    )
+
+    # Let the Agent always choose "Retriever" as the Tool with "" as input
+    def mock_extract_tool_name_and_tool_input(self, pred: str) -> Tuple[str, str]:
+        return "Retriever", ""
+
+    monkeypatch.setattr(Agent, "_extract_tool_name_and_tool_input", mock_extract_tool_name_and_tool_input)
+
+    # Using max_iterations as specified in the Agent's init method
+    with caplog.at_level(logging.WARN, logger="haystack.agents"):
+        result = agent.run("Where does Christelle live?")
+    assert result["answers"] == [Answer(answer="", type="generative")]
+    assert "Maximum number of iterations (1) reached" in caplog.text
+
+    # Setting max_iterations in the Agent's run method
+    with caplog.at_level(logging.WARN, logger="haystack.agents"):
+        result = agent.run("Where does Christelle live?", max_iterations=0)
+    assert result["answers"] == [Answer(answer="", type="generative")]
+    assert "Maximum number of iterations (0) reached" in caplog.text
```

Comment:
Maybe we should raise an exception when a user sets `max_iterations` to 0? I don't see the reason why that should be possible?

Reply:
There is no good reason for setting `max_iterations` to 0 or 1, that's correct. I will change the test to use 3 and 2 instead of 1 and 0. Not sure whether it's worth checking and raising an exception though. I'll need to think about it.

Reply:
We'll raise an error now if `max_iterations` is smaller than 2.